crime-evidence/src/main/java/uk/gov/justice/laa/crime/evidence/service/IncomeEvidenceValidationService.java

```diff
@@ -54,9 +62,55 @@ public void checkExtraEvidenceDescriptions(List<ApiIncomeEvidence> incomeEvidenc
             .filter(apiIncomeEvidence -> EXTRA_EVIDENCES.contains(apiIncomeEvidence.getEvidenceType()))
             .forEach(apiIncomeEvidence -> {
                     if (StringUtils.isBlank(apiIncomeEvidence.getDescription())) {
-                        throw new IllegalArgumentException(MISSING_OTHER_EVIDENCE_DESCRIPTION);
+                        throw new CrimeEvidenceDataException(MISSING_OTHER_EVIDENCE_DESCRIPTION);
                     }
                 }
             );
     }
+
+    public void validateUpliftDates(UpdateEvidenceDTO updateEvidenceDTO, boolean allEvidenceReceived) {
+        LocalDate upliftAppliedDate = updateEvidenceDTO.getUpliftAppliedDate();
+        LocalDate upliftRemovedDate = updateEvidenceDTO.getUpliftRemovedDate();
+        LocalDate oldUpliftAppliedDate = updateEvidenceDTO.getOldUpliftAppliedDate();
+        LocalDate oldUpliftRemovedDate = updateEvidenceDTO.getOldUpliftRemovedDate();
+        LocalDate currentDate = LocalDate.now();
+
+        if (upliftRemovedDate != null) {
+            if (oldUpliftAppliedDate == null) {
+                throw new CrimeEvidenceDataException(CANNOT_SET_UPLIFT_REMOVED_DATE_WHEN_NO_UPLIFT_APPLIED);
+            }
+
+            if (!upliftRemovedDate.equals(oldUpliftRemovedDate) && !upliftRemovedDate.equals(currentDate)) {
+                throw new CrimeEvidenceDataException(MUST_SET_UPLIFT_REMOVED_DATE_TO_TODAY);
+            }
+        }
+
+        if (oldUpliftRemovedDate != null) {
+            if (upliftRemovedDate == null) {
+                throw new CrimeEvidenceDataException(CANNOT_CLEAR_UPLIFT_REMOVED_DATE);
+            } else if (!upliftRemovedDate.equals(oldUpliftRemovedDate)) {
+                throw new CrimeEvidenceDataException(CANNOT_MODIFY_UPLIFT_REMOVED_DATE);
+            }
+        }
+
+        if (upliftAppliedDate != null) {
+            if (oldUpliftAppliedDate == null && allEvidenceReceived) {
+                throw new CrimeEvidenceDataException(CANNOT_APPLY_UPLIFT_IF_NO_OUTSTANDING_EVIDENCE_REQUIRED);
+            }
+
+            if (!upliftAppliedDate.equals(oldUpliftAppliedDate) && !upliftAppliedDate.equals(currentDate)) {
+                throw new CrimeEvidenceDataException(MUST_SET_UPLIFT_DATE_TO_TODAY);
+            }
+        }
+
+        if (oldUpliftAppliedDate != null) {
+            if (upliftAppliedDate == null) {
+                throw new CrimeEvidenceDataException(CANNOT_CLEAR_UPLIFT_APPLIED_DATE);
+            }
+            if (oldUpliftRemovedDate == null && !upliftAppliedDate.equals(oldUpliftAppliedDate)) {
+                throw new CrimeEvidenceDataException(CANNOT_SET_NEW_UPLIFT_DATE);
+            }
+        }
```

Comment:
I find this quite hard to understand what's going on, is it worth splitting out these checks into separate methods which meaningfully describe the validation checks that are taking place?

Reply:
Thanks for refactoring this 👍